sdk/core/corehttp/corehttp/streaming/_stream.py

```diff
@@ -0,0 +1,173 @@
+#
+# The MIT License (MIT)
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the ""Software""), to
+# deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+# sell copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+
+from types import TracebackType
+from typing import Iterator, AsyncIterator, TypeVar, Callable, Any, Optional, Type
+
+from typing_extensions import Self
+
+from ..rest import HttpResponse, AsyncHttpResponse
+from .decoders import JSONLDecoder, AsyncJSONLDecoder, StreamDecoder, AsyncStreamDecoder
+
+
+ReturnType = TypeVar("ReturnType")
+
+
+class Stream(Iterator[ReturnType]):
+    """Stream class for streaming JSONL or Server-Sent Events (SSE).
+
+    :keyword response: The response object.
+    :paramtype response: ~corehttp.rest.HttpResponse
+    :keyword deserialization_callback: A callback that takes JSON and returns a deserialized object.
+    :paramtype deserialization_callback: Callable[[Any], ReturnType]
+    :keyword decoder: An optional decoder to use for the stream. If not provided, the content-type
+        of the response will be used to determine the decoder.
+    :paramtype decoder: Optional[~corehttp.streaming.decoders.StreamDecoder]
+    :paramtype terminal_event: Optional[str]
+    :keyword terminal_event: A terminal event that indicates the end of the SSE stream.
+    :paramtype terminal_event: Optional[str]
+    """
+
+    def __init__(
+        self,
+        *,
+        response: HttpResponse,
+        deserialization_callback: Callable[[Any], ReturnType],
+        decoder: Optional[StreamDecoder] = None,
+        terminal_event: Optional[str] = None,
+    ) -> None:
+        self._response = response
+        self._deserialization_callback = deserialization_callback
+        self._terminal_event = terminal_event
+        self._iterator = self._iter_results()
+
+        if decoder is not None:
+            self._decoder = decoder
+        elif self._response.headers.get("Content-Type", "").lower() == "application/jsonl":
```

Comment:
This would fail if the service provided any [parameters](https://developer.mozilla.org/en-US/docs/Web/HTTP/MIME_types#structure_of_a_mime_type) in the MIME type. Should we make the decoder a required parameter?

Reply:
You're right - making it required. The generated code should know which flavor of decoder to pass, anyway.